ouroboros-network/src/Ouroboros/Network/Server/Socket.hs

```diff
@@ -0,0 +1,223 @@
+      (addr, channel) <- io
+      pure (addr, channel, pure ())
+  }
+
+type StatusVar st = STM.TVar st
+
+-- | What to do with a new connection: reject it and give a new state, or
+-- accept it and give a new state with a continuation to run against the
+-- resulting channel.
+-- See also `CompleteConnection`, which is run for every connection when it finishes, and
+-- can also update the state.
+data HandleConnection channel st r where
+  Reject :: !st -> HandleConnection channel st r
+  Accept :: !st -> !(channel -> IO r) -> HandleConnection channel st r
+
+-- | What to do on a new connection: accept and run this `IO`, or reject.
+type BeginConnection addr channel st r = addr -> st -> STM (HandleConnection channel st r)
+
+-- | How to update state when a connection finishes. Can use `throwSTM` to
+-- terminate the server.
+type CompleteConnection st r = Either SomeException r -> st -> STM st
+
+-- | Given a current state, `retry` unless you want to stop the server.
+-- When this transaction returns, any running threads spawned by the server
+-- will be killed.
+--
+-- It's possible that a connection is accepted after the main thread
+-- returns, but before the server stops. In that case, it will be killed, and
+-- the `CompleteConnection` will not run against it.
+type Main st t = st -> STM t
+
+-- | To avoid repeatedly blocking on the set of all running threads (a
+-- potentially very large STM transaction) the results come in by way of a
+-- `TQueue`. Using a queue rather than, say, a `TMVar`, also finesses a
+-- potential deadlock when shutting down the server and killing spawned threads:
+-- the server can stop pulling from the queue, without causing the child
+-- threads to hang attempting to write to it.
+type ResultQ r = STM.TQueue (Result r)
+
+-- | The product of a spawned thread. We catch all (even async) exceptions.
+data Result r = Result
+  { resultThread :: !(Async ())
+  , resultValue  :: !(Either SomeException r)
+  }
+
+-- | The set of all spawned threads. Used for waiting or cancelling them when
+-- the server shuts down.
+type ThreadsVar = STM.TVar (Set (Async ()))
+
+-- | The action runs inside `try`, and when it finishes, puts its result
+-- into the `ResultQ`. Takes care of inserting/deleting from the `ThreadsVar`.
+--
+-- Async exceptions are masked to ensure that if the thread is spawned, it
+-- always gets into the `ThreadsVar`. Exceptions are unmasked in the
+-- spawned thread.
+spawnOne
+  :: ResultQ r
+  -> ThreadsVar
+  -> IO r
+  -> IO ()
+spawnOne resQ threadsVar io = mask_ $ do
+  rec let threadAction = \unmask -> do
+            val <- try (unmask io)
+            -- No matter what the exception, async or sync, this will not
+            -- deadlock, since we use a `TQueue`. If the server kills its
+            -- children, and stops clearing the queue, it will be collected
+            -- shortly thereafter, so no problem.
+            STM.atomically $ STM.writeTQueue resQ (Result thread val)
```

Comment:
@avieth if you return the `val` you will not need to include it in `Result`; or even better, just remove the `thread` since there's no need to use it if the `val` is already available (this way `rec` will not be needed as well).

Reply:
The `thread` is needed so that we can remove it from the set of running threads (see `connectionTx`).

Reply:
I don't understand what you mean by "if you return the `val`". All I do with the `val` is put it into the `resQ`, so that `connectionTx` can grab it later.

Reply:
I mean that `thread :: Async (Either SomeException t)` will give access to it, wehn you add `return val` at the end:
```
rec let threadAction = \unmask -> do
            val <- try (unmask io)
            STM.atomically $ STM.writeTQueue resQ (Result thread)
            return val
```

Reply:
But maybe the semantics is a bit wired: you'll need to call `await` on something that one knows that has already terminated.

Reply:
True, we could just dump the async into the `resQ` without the `val` and then `await` it, knowing it will not block. I think that would be exactly the same.

Reply:
We would still need the recursive do, because the thread itself must put the `Async` of itself into the queue.

Reply:
On the other hand: it's probably just `ThreadId` that is needed by `connectionTx`; could we use it instead of `Async`? there there would be no `rec` and `Result` could contain just the value.

Reply:
Yes, _technically_ only the `ThreadId` is needed. But, the set of running threads should be `Async`s, not `ThreadId`s, so that we can `cancel` them, which is different than simply `throwTo` on the `ThreadId` (it also waits for the thread to die). Since the set must contain `Async`s, and those have `Ord` instances on `ThreadId`, it's just simpler to use the `Async` here as well. Otherwise, it's awkward and brittle to remove from the set: we'd have to make an `Async ()` from just a `ThreadId`, which could technically work by using `undefined` or `retry` for the await transaction, but I'd rather not.

Reply:
> Then there would be no rec and Result could contain just the value.

Result needs to contain enough information to remove the thread from the set of running threads.

Reply:
ok, I forgot that we need  access to `cancel` as well.